src/rtutils.c

```diff
@@ -777,6 +812,55 @@ static int jl_static_is_function_(jl_datatype_t *vt) JL_NOTSAFEPOINT {
     return 0;
 }
 
+static size_t jl_static_show_float(JL_STREAM *out, double v,
+                                   jl_datatype_t *vt) JL_NOTSAFEPOINT
+{
+    size_t n = 0;
+    // TODO: non-canonical NaNs do not round-trip
+    // TOOD: BFloat16
+    const char *size_suffix = vt == jl_float16_type ? "16" :
+                              vt == jl_float32_type ? "32" :
+                                                      "";
+    // Requires minimum 1 (sign) + 17 (sig) + 1 (dot) + 5 ("e-123") + 1 (null)
+    char buf[32];
+    // Base B significand digits required to print n base-b significand bits
+    // (including leading 1):  N = 2 + floor(n/log(b, B))
+    //   Float16   5
+    //   Float32   9
+    //   Float64  17
+    // REF: https://dl.acm.org/doi/pdf/10.1145/93542.93559
+    if (isnan(v)) {
+        n += jl_printf(out, "NaN%s", size_suffix);
+    }
+    else if (isinf(v)) {
+        n += jl_printf(out, "%sInf%s", v < 0 ? "-" : "", size_suffix);
+    }
+    else if (vt == jl_float64_type) {
+        size_t m = snprintf(buf, sizeof buf, "%.17g", v);
+        jl_uv_puts(out, buf, m);
+        n += m;
+        // If there is no decimal point or exponent, disambiguate
+        if (!strpbrk(buf, ".e"))
+            jl_printf(out, ".0");
```

Comment:
While clever, would it be better to use the standard flag `#` for this to preserve all trailing zeros (rather than to imply the trailing digit is `.0` which may not be entirely true given final digit rounding, though I think it would have to round-trip the same since 17 digits is full accuracy)
```suggestion
        n += jl_printf(buf, sizeof buf, "%#.17g", v);
```

Reply:
I assume that later printed as `Float16(-0)`, no? (though that makes it `isequal 0.0`, not `-0.0`)